File: src/rust/day04.rs
#[derive(Debug)]
pub struct Card {
    pub id: i32,
    winning_numbers: Vec<i32>,
    guess_numbers: Vec<i32>,
}

impl Card {
    pub fn from_string(s: &str) -> Card {
        let parts: Vec<&str> = s.splitn(2, ':').collect();
        let card = parts[0];
        let numbers = parts[1];

        let id: i32 = card
            .split_whitespace()
            .nth(1)
            .expect("Error parsing Card ID")
            .parse()
            .expect("Error converting Card ID to i32");

        let number_parts: Vec<&str> = numbers.splitn(2, '|').collect();
        let winning_numbers: Vec<i32> = number_parts[0]
            .split_whitespace()
            .map(|n| n.parse().expect("Error converting winning number to i32"))
            .collect();
        let guess_numbers: Vec<i32> = number_parts[1]
            .split_whitespace()
            .map(|n| n.parse().expect("Error converting guess number to i32"))
            .collect();
        Card {
            id,
            winning_numbers,
            guess_numbers,
        }
    }

    pub fn score(&self) -> i32 {
        let mut result: i32 = 0;
        for n in &self.guess_numbers {
            if self.winning_numbers.contains(&n) {
                if result == 0 {
                    result = 1;
                } else {
                    result = result * 2
                }
            }
        }
        result
    }

    pub fn copies(&self) -> Vec<i32> {
        let copy_range = self
            .guess_numbers
            .iter()
            .filter(|n| self.winning_numbers.contains(n))
            .count();
        let start = self.id + 1;
        let stop = start + copy_range as i32;
        (start..stop).collect()
    }
}

File: python/aoc_2023/day04.py
from dataclasses import dataclass
from typing import List


@dataclass
class Card:
    id: int
    winning_numbers: List[int]
    guess_numbers: List[int]

    @classmethod
    def from_string(cls, s: str) -> "Card":
        card, numbers = s.split(":", 1)
        id = int(card.split()[1])
        (left, right) = numbers.split("|", 1)
        winning_numbers = left.split()
        guess_numbers = right.split()
        return cls(id=id, winning_numbers=winning_numbers, guess_numbers=guess_numbers)

    def score(self) -> int:
        score = 0
        for n in self.guess_numbers:
            if n in self.winning_numbers:
                if not score:
                    score = 1
                else:
                    score *= 2
        return score

    def copies(self) -> List[int]:
        copy_range = len([n for n in self.guess_numbers if n in self.winning_numbers])
        start = self.id + 1
        stop = start + copy_range
        return list(range(start, stop))
